docs/arcadia-users-group/20240206-intro-to-formatting-and-linting/lesson.md

```diff
@@ -0,0 +1,433 @@
+
+### Documentation conventions
+In the context of code style guides, "documentation" refers to human-readable text that is embedded in the source code to explain what the code does and how it works. It is very important to define standards and conventions for documentation because it is a major way--and sometimes the only way--to ensure that code is readily understandable by others. In Python, documentation takes two forms: comments that can appear anywhere in the code and docstrings that accompany modules, classes, and functions.
+
+#### Comments
+Comments are human-readable lines of text that can appear anywhere in the source code. In Python, they are denoted by the `#` character. Although they are ignored by the Python interpreter and might seem like an area where "anything goes", it is important to adhere to strict standards of grammar and style when writing comments, since they are often the only way to explain the purpose of a particular line or block of code.
+
+In particular, comments should be complete sentences with proper punctuation. They should be written in full English sentences and should be grammatically correct. Importantly, they should end with periods; this is not only grammatically correct but is also the only way to indicate to the reader that the comment is complete and was not accidentally truncated (or never completely written in the first place). 
+
+Most importantly, comments should generally be used to explain *why* a particular line or block of code is doing what it is doing, not *what* it is doing or *how* it is doing it (usually, this should be apparent from the code itself). 
+
+Here is an example of a good comment that explains *why* a CSV file is loaded with a particular set of parameters:
+```python
+# We can assume missing values are always represented by the string 'NaN' in the CSV file,
+# so the parameter 'na_values' is set to 'NaN'.
+df = pd.read_csv('data.csv', na_values='NaN', keep_default_na=False)
+```
+
+Here is another example of a good comment that explains a non-obvious implementation detail when computing the nth Fibonacci number:
+```python
+def calculate_fibonacci(n):
+    """
+    Calculate the nth Fibonacci number using naive recursion.
+    """
+    # By definition, the Fibonacci sequence starts with zero and one.
+    if n <= 0:
+        return 0
+    elif n == 1:
+        return 1
+    else:
+        return fibonacci(n - 1) + fibonacci(n - 2)
+```
+
+#### Things comments should *not* be used for
+There are two major ways in which comments are commonly misused. Note that avoiding these misuses is not specific to Python but is a general best practice that applies to all programming languages. The first is that, as we alluded to above, comments should not be redundant with the code. They should not simply restate what the code is doing and they should not include information that is apparent or readily inferred from the code itself. Here are some examples of such redundant comments:
```

Comment:
Could be worth calling out here that this is the opposite of what ChatGPT does, which makes ChatGPT code very annoying to work with.